2019/pkg/intcode/intcode.go

```diff
@@ -51,91 +66,67 @@ type Program struct {
 	output         []int
 }
 
-func NewProgram(input []int) *Program {
+func NewProgram(instructions []int) *Program {
 	return &Program{
-		instructions: input,
+		instructions: instructions,
 		output:       []int{},
 	}
 }
 
-func (p *Program) Run(input int) (int, error) {
-	for p.programCounter < len(p.instructions) {
-		op, addressingModes, err := parseOpCodeAddressMode(p.instructions[p.programCounter])
+func (p *Program) Run(input int) error {
+	for {
+		opcode, parameterIndexes, err := p.Parse()
```

Comment:
A note about performance. For each step of the program you end up allocating a slice for parameters. One thing you can try at some point is to see how much you can make the interpreter faster... however, you can leave that for later.

Reply:
The slice of parameters at this point only has at most 3 items in it so that doesn't seem like a huge performance hit.  But what are some ideas to improve the performance?  Reuse the same slice?

Reply:
Reusing the slice is one way the other is to avoid slices altogether and get the exact number arguments where commands are handled.

I recommend benchmarking some custom program with a loop that's iterated 100k times and some basic math.

Reply:
I don't understand what you mean by the benchmark recommendation. Are you saying to benchmark allocating a new slice in a loop where the loop runs 100k times?